internal/proto6/request_type_test.go

```diff
@@ -0,0 +1,155 @@
+package proto6
+
+import (
+	"context"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+func TestRequestType(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		req         *tfprotov6.ApplyResourceChangeRequest
+		typ         tftypes.Type
+		isCreate    bool
+		isUpdate    bool
+		isDestroy   bool
+		expectedErr string
+	}
+
+	reqType := tftypes.Object{
+		AttributeTypes: map[string]tftypes.Type{
+			"a": tftypes.String,
+			"b": tftypes.Number,
+			"c": tftypes.Bool,
+		},
+	}
+
+	setDV, err := tfprotov6.NewDynamicValue(reqType, tftypes.NewValue(reqType, map[string]tftypes.Value{
+		"a": tftypes.NewValue(tftypes.String, "hello, world"),
+		"b": tftypes.NewValue(tftypes.Number, 123),
+		"c": tftypes.NewValue(tftypes.Bool, true),
+	}))
+	if err != nil {
+		t.Errorf("Unexpected error creating set dynamic value: %s", err)
+		return
+	}
+
+	nullDV, err := tfprotov6.NewDynamicValue(reqType, tftypes.NewValue(reqType, nil))
+	if err != nil {
+		t.Errorf("Unexpected error creating null dynamic value: %s", err)
+		return
+	}
+
+	tests := map[string]testCase{
+		"null-prior-state-set-planned-state": {
+			req: &tfprotov6.ApplyResourceChangeRequest{
+				TypeName:     "foo",
+				PriorState:   &nullDV,
+				PlannedState: &setDV,
+			},
+			typ:       reqType,
+			isCreate:  true,
+			isUpdate:  false,
+			isDestroy: false,
+		},
+		"set-prior-state-set-planned-state": {
+			req: &tfprotov6.ApplyResourceChangeRequest{
+				TypeName:     "foo",
+				PriorState:   &setDV,
+				PlannedState: &setDV,
+			},
+			typ:       reqType,
+			isCreate:  false,
+			isUpdate:  true,
+			isDestroy: false,
+		},
+		"set-prior-state-null-planned-state": {
+			req: &tfprotov6.ApplyResourceChangeRequest{
+				TypeName:     "foo",
+				PriorState:   &setDV,
+				PlannedState: &nullDV,
+			},
+			typ:       reqType,
+			isCreate:  false,
+			isUpdate:  false,
+			isDestroy: true,
+		},
+	}
+
+	for name, tc := range tests {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			got, err := IsCreate(context.Background(), tc.req, tc.typ)
+			if err != nil {
+				if tc.expectedErr == "" {
```

Comment:
Yeah, sorry, I mostly just wanted to have it so adding tests that provoke errors is easy. I know YAGNI, but tests are one of the places where I like to try and predict future needs, because making adding future cases easy means more comprehensive testing, usually. So when I do table-driven tests, I try to make sure all the non-context.Context arguments can be specified and all the returned values can be checked. (I'm not always perfect at it, but it's a general habit at this point.)